components/overlayTrigger/overlayTrigger.js

```diff
@@ -0,0 +1,155 @@
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+
+export default class OverlayTrigger extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      active: false,
+    };
+  }
+
+  linkChild = (ref) => {
+    this.elem = ref;
+  }
+
+  componentDidMount() {
+    if (this.props.triggerAction !== 'click') {
+      return;
+    }
+
+    this.addOutsideClickListener();
+  }
+
+  componentWillUnmount() {
+    if (this.props.triggerAction !== 'click') {
+      return;
+    }
+
+    this.removeOutsideClickListener();
+  }
+
+  addOutsideClickListener() {
+    document.body.addEventListener('click', this.handleOutsideClick);
+  }
+
+  removeOutsideClickListener() {
+    document.body.removeEventListener('click', this.handleOutsideClick);
+  }
+
+  handleOutsideClick = (e) => {
+    if (this.elem.contains(e.target)) {
+      return;
+    }
+
+    this.state.active && this.hideElement();
+  }
+
+  toggleElement = () => {
+    this.state.active
+      ? this.hideElement()
+      : this.showElement();
+  }
+
+  showElement() {
+    const { onElementShow } = this.props;
+
+    this.setState({ active: true });
+
+    onElementShow && onElementShow();
+  }
+
+  hideElement() {
+    const { onElementHide } = this.props;
+
+    this.setState({ active: false });
+
+    onElementHide && onElementHide();
+  }
+
+  getOnClickTarget() {
+    const { children } = this.props;
+
+    return (
+      <div onClick={this.toggleElement}>
+        {children}
+      </div>
+    );
+  }
+
+  getOnHoverTarget() {
+    const { children } = this.props;
+
+    return (
+      <div onMouseOut={this.toggleElement} onMouseOver={this.toggleElement}>
+        {children}
+      </div>
+    );
+  }
+
+  handleCloseButtonClick = () => {
+    this.hideElement();
+  }
+
+  render() {
+    const { triggerAction } = this.props;
+
+    const actions = {
+      click: this.getOnClickTarget,
+      hover: this.getOnHoverTarget,
+    };
+
+    const targetElement = actions[triggerAction].call(this);
```

Comment:
This is the same as binding.

If you don't want to set the `this.getOnClickTarget = this.getOnClickTarget.bind(this);` in the `constructor`, at least do the `getOnClickTarget = () => { /* ... */ }` to not reset the context on render ;)

Reply:
done